### PR TITLE
Try reducing allocation in stream processing

### PR DESCRIPTION
**Pull request by jbachorik**
This change is introducing a `RequestBody` subclass which can do direct stream-to-stream compression without going through the intermediary state of reading in all data into a byte array. This should lead to decreased allocation pressure coming from the profile uploads.

Thanks to handling the compression in Okio specific manner the generic `StreamUtils` class which was providing compression helper methods was removed.

Because the upload data size is only known after the input stream has been fully compressed it is not possible to send it in the header and/or log before starting the upload request. Instead, the request is using chunked transfer with the request size set to `-1` (== unknown) and the request debug logging is done in the request callback once the request has finished/failed.

### CODE REVIEW

**Comment by richardstartin** [comment]
Is there a good reason not to write this out directly to the HTTP request's output stream?

**Comment by jbachorik** [comment]
IIRC, the reason was that we were not able to provide the actual body content length if doing directly stream-to-stream and we are using that for metrics. Also, but here I might be wrong, I can't find a way to easily attach a stream to an okhttp request.

**Comment by richardstartin** [comment]
It's possible to extend `RequestBody` which lets you write to `ByteSink`, so you can do the compression into a stream which writes to the `ByteSink` there. This would preclude setting `Content-Length` though. Would it be enough to record how many bytes were written/read for metrics, depending on whether these metrics are server or client side?

**Comment by jbachorik** [comment]
> This would preclude setting Content-Length though

I will check whether this would cause any problems for backend.

**Comment by richardstartin** [comment]
@jbachorik do you think this is worth pursuing or should we close it?

**Comment by jbachorik** [comment]
@richardstartin I would give it one more try. I have an implementation that does what is expected - just want to run the tests on CircleCI (my MBP 13" is getting killed by the tests :/) and then some finishing touches (perhaps byte counting and sizes reporting in logs) and will be ready for the real PR.

File: dd-java-agent/agent-profiling/profiling-uploader/src/main/java/com/datadog/profiling/uploader/CompressingRequestBody.java [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+                      }
+                    }));
+    BufferedSink sink = Okio.buffer(Okio.sink(sinkStream));
+    try (Source source = Okio.buffer(Okio.source(inputStream))) {
+      sink.writeAll(source);
+    }
+    // a bit of cargo-culting to make sure that all writes have really-really been flushed
+    sink.emit();
+    sink.flush();
+    sinkStream.close();
```
Comment (jpbempel) [inline dd-java-agent/agent-profiling/profiling-uploader/src/main/java/com/datadog/profiling/uploader/CompressingRequestBody.java:225]: try-with-resources?

**Review by jpbempel — APPROVED** [review]
LGTM

**Review by richardstartin — APPROVED** [review]
👍🏻

**Pull request merged**